utils/logalloc.cc

```diff
@@ -361,7 +365,74 @@ static thread_local tracker& tracker_instance = get_tracker_instance();
 
 using clock = std::chrono::steady_clock;
 
+class background_reclaimer {
+    scheduling_group _sg;
+    noncopyable_function<void (size_t target)> _reclaim;
+    timer<lowres_clock> _adjust_shares_timer;
+    // If engaged, main loop is not running, set_value() to wake it.
+    promise<>* _main_loop_wait = nullptr;
+    future<> _done;
+    bool _stopping = false;
+    static constexpr size_t free_memory_threshold = 60'000'000;
+private:
+    bool have_work() const {
+#ifndef SEASTAR_DEFAULT_ALLOCATOR
+        return memory::stats().free_memory() < free_memory_threshold;
+#else
+        return false;
+#endif
+    }
+    void main_loop_wake() {
+        llogger.debug("background_reclaimer::main_loop_wake: waking {}", bool(_main_loop_wait));
+        if (_main_loop_wait) {
+            _main_loop_wait->set_value();
+            _main_loop_wait = nullptr;
+        }
+    }
+    future<> main_loop() {
+        llogger.debug("background_reclaimer::main_loop: entry");
+        while (true) {
+            while (!_stopping && !have_work()) {
+                promise<> wait;
+                _main_loop_wait = &wait;
+                llogger.trace("background_reclaimer::main_loop: sleep");
+                co_await wait.get_future();
+                llogger.trace("background_reclaimer::main_loop: awakened");
+                _main_loop_wait = nullptr;
+            }
+            if (_stopping) {
+                break;
+            }
+            _reclaim(free_memory_threshold - memory::stats().free_memory());
+            co_await make_ready_future<>();
+        }
+        llogger.debug("background_reclaimer::main_loop: exit");
+    }
+    void adjust_shares() {
+        if (_main_loop_wait && have_work()) {
+            _sg.set_shares(1 + (1000 * (free_memory_threshold - memory::stats().free_memory())) / free_memory_threshold);
+            main_loop_wake();
+        }
+    }
+public:
+    explicit background_reclaimer(scheduling_group sg, noncopyable_function<void (size_t target)> reclaim)
+            : _sg(sg)
+            , _reclaim(std::move(reclaim))
+            , _adjust_shares_timer(_sg, [this] { adjust_shares(); })
+            , _done(with_scheduling_group(_sg, [this] { return main_loop(); })) {
+        if (sg != default_scheduling_group()) {
+            _adjust_shares_timer.arm_periodic(50ms);
```

Comment:
Maybe it makes sense to make this a poller so that it wakes with no delay and has arbitrary logic.

Reply:
Pollers are not exposed outside seastar.

As I mentioned elsewhere, one option is to have a one-shot notification when memory drops below a threshold. memory.cc already does a compare, so the cost is zero. After we're back over the threshold, we reinstate the notification.